### PR TITLE
fix - testing role managment 

### PR DESCRIPTION
**Pull request by bellaabdelouahab**
related issue and discussion #49 #48 #47 

- [X] add route /me for user crud that are protected 
- [X] a user shouldn't be able to updated roles
- [X] if user tries to updated password we replay with ( 400, This route is not for password updates. Please use /updateMyPassword )
- [X] admin can update roles of users and admins but not supper admin. 

> note : we can add a restriction called updated admin, to prevent the fight between admins 😂

- [X] supper admin can updated any role of any user except his. 
- [X] return a valid message on jwt expired, as for now it returns 500
- [X] separate the routes login and sign up from the rest as they are the only unprotected routes ( suggestion : put them in app.js)
- [X] auto generate swagger docs, but still needs description for post requests
the roles management still needs more testing

### CODE REVIEW

**Review by muttaqin1 — APPROVED** [review]
good job👌

**Comment by Khalid1G** [comment]
Just a small change
> if user tries to updated password we replay with ( 400, This route is not for password updates. Please use /updateMyPassword )

You have too return 403 instead of 400

**Comment by bellaabdelouahab** [comment]
But wait i haven't merged my changes yet @muttaqin1

**Comment by muttaqin1** [comment]
okay

File: backend-app/controllers/authController.js [hunk 1]
```diff
@@ -157,16 +157,28 @@ exports.protect = async (req, res, next) => {
     req.user = user;
```
Comment (muttaqin1) [inline backend-app/controllers/authController.js:157]: i guess req.user includes the password so the **getMe** controller should make the password property null then send the response.
@bellaabdelouahab

**Review by muttaqin1 — CHANGES_REQUESTED** [review]

File: backend-app/controllers/userController.js [hunk 2]
```diff
@@ -18,6 +29,40 @@ exports.deleteMe = async (req, res, next) => {
...
+        // 1) Create error if user POSTs password data
+        if (req.body.password || req.body.passwordConfirm) {
+            return next(new AppError(400, 'fail', 'This route is not for password updates. Please use /updateMyPassword'), req, res, next);
+        }
+        // create error if user tries to update role
+        if (req.body.roles) {
+            return next(new AppError(400, 'fail', 'This route is not for role updates. Please use /updateRole'), req, res, next);
+        }
+        // 2) Filtered out unwanted fields names that are not allowed to be updated
+        const filteredBody = Object.keys(req.body).filter(el => el !== 'name' && el !== 'email');
```
Comment (Khalid1G) [inline backend-app/controllers/userController.js:44]: I think you should allow user to change his name

File: backend-app/controllers/userController.js [hunk 3]
```diff
@@ -18,6 +29,40 @@ exports.deleteMe = async (req, res, next) => {
     }
 };
 
+
+exports.updateMe = async (req, res, next) => {
+    try {
+        // 1) Create error if user POSTs password data
+        if (req.body.password || req.body.passwordConfirm) {
+            return next(new AppError(400, 'fail', 'This route is not for password updates. Please use /updateMyPassword'), req, res, next);
```
Comment (Khalid1G) [inline backend-app/controllers/userController.js:37]: the status code is 403 not 400

**Review by Khalid1G — COMMENTED** [review]
allow the user to change his name, also change the status code from 400 to 403 in base code and docs

File: backend-app/controllers/authController.js (hunk 1 above)
Comment (bellaabdelouahab) [reply backend-app/controllers/authController.js:157]: ah yes i will fix it in next pr @muttaqin1

File: backend-app/controllers/userController.js (hunk 2 above)
Comment (bellaabdelouahab) [reply backend-app/controllers/userController.js:44]: sure i ll fix it in next pr

File: backend-app/controllers/userController.js (hunk 3 above)
Comment (bellaabdelouahab) [reply backend-app/controllers/userController.js:37]: you right.

**Review by Khalid1G — APPROVED** [review]

**Pull request merged**